File: solution/200. Number of Islands/200_02.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        m = grid.size(), n = grid[0].size();
        
        int res = 0;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                res += grid[i][j] - '0';
                dfs(grid, i, j);
            }
        }
        return res;
    }
    
private:
    int m, n;
    const vector<int> dirs = {0, -1, 0, 1, 0};
    
    void dfs(vector<vector<char>>& grid, int y, int x) {
        if (y < 0 || y >= m || x < 0 || x >= n || grid[y][x] == '0') return;
        
        grid[y][x] = '0';
        for (int k = 0; k < 4; ++k)
            dfs(grid, y + dirs[k], x + dirs[k + 1]);
    }
    
};

File: solution/200. Number of Islands/200_02.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        dirs = [0, 1, 0, -1, 0]
        def dfs(i, j):
            if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] != '1': return
            grid[i][j] = '0'
            for k in range(4):
                dfs(i + dirs[k], j + dirs[k + 1])
        
        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    res += 1
                    dfs(i, j)
        return res
